be/src/cloud/cloud_snapshot_loader.cpp

```diff
@@ -0,0 +1,234 @@
+                                         const TNetworkAddress& broker_addr,
+                                         const std::map<std::string, std::string>& broker_prop)
+        : BaseSnapshotLoader(env, job_id, task_id, broker_addr, broker_prop),
+          _engine(engine) {};
+
+Status CloudSnapshotLoader::init(TStorageBackendType::type type, const std::string& location, std::string vault_id) {
+    RETURN_IF_ERROR(BaseSnapshotLoader::init(type, location));
+    _storage_resource = _engine.get_storage_resource(vault_id);
+    if (!_storage_resource) {
+        return Status::InternalError("vault id not found, vault id {}", vault_id);
+    }
+    return Status::OK();
+}
+
+io::RemoteFileSystemSPtr CloudSnapshotLoader::storage_fs() {
+    return _storage_resource->fs;
+}
+
+Status CloudSnapshotLoader::upload(const std::map<std::string, std::string>& src_to_dest_path,
+                                   std::map<int64_t, std::vector<std::string>>* tablet_files) {
+    return Status::NotSupported("upload not supported");
+}
+
+Status CloudSnapshotLoader::download(const std::map<std::string, std::string>& src_to_dest_path,
+                                     std::vector<int64_t>* downloaded_tablet_ids) {
+    if (!_remote_fs || !_storage_resource) {
+        return Status::InternalError("Storage backend not initialized.");
+    }
+
+    LOG(INFO) << "begin to transfer snapshot files. num: " << src_to_dest_path.size()
+              << ", broker addr: " << _broker_addr << ", job: " << _job_id
+              << ", task id: " << _task_id;
+
+    // check if job has already been cancelled
+    int tmp_counter = 1;
+    RETURN_IF_ERROR(_report_every(0, &tmp_counter, 0, 0, TTaskType::type::DOWNLOAD));
+
+    Status status = Status::OK();
+
+    // 1. for each src path, transfer files to target path
+    int report_counter = 0;
+    int total_num = src_to_dest_path.size();
+    int finished_num = 0;
+    for (const auto& iter : src_to_dest_path) {
+        const std::string& remote_path = iter.first;
+        const std::string& tablet_str = iter.second;
+        int64_t target_tablet_id = -1;
+        try {
+            target_tablet_id = std::stoll(tablet_str);
+        } catch (std::exception& e) {
+            return Status::InternalError("failed to parse target tablet id {}, {}", tablet_str, e.what());
+        }
+        const std::string target_path = _storage_resource->remote_tablet_path(target_tablet_id);
+
+        // 1.1. check target path not exists
+        bool target_path_exist = false;
+        if (!storage_fs()->exists(target_path, &target_path_exist).ok() ||
+                target_path_exist) {
+            std::stringstream ss;
+            ss << "failed to download snapshot files, target path already exists: " << target_path;
+            LOG(WARNING) << ss.str();
+            return Status::InternalError(ss.str());
+        }
```

Comment:
If a retry happens, then it fails?

Reply:
If one download task failed, the restore job cancelled. The restore job cannot restart again.
And new restore job should carry a new tablet id.